File: Trabajo_3/Ejercicios/Ejercicio_cpp/Ejercicio_3.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Asignatura;

class Estudiante {
private:
    string nombre;
    string direccion;
    string numeroMatricula;
    vector<Asignatura*> asignaturasMatriculadas;
    vector<int> notas;
    vector<int> faltasAsistencia;

public:
    Estudiante(string _nombre, string _direccion, string _numeroMatricula) : nombre(_nombre), direccion(_direccion), numeroMatricula(_numeroMatricula) {}

    // Métodos para acceder y modificar los atributos
    string getNombre() const { return nombre; }
    string getDireccion() const { return direccion; }
    string getNumeroMatricula() const { return numeroMatricula; }
    void matricularAsignatura(Asignatura* asignatura) { asignaturasMatriculadas.push_back(asignatura); }
    void agregarNota(int nota) { notas.push_back(nota); }
    void registrarFaltaAsistencia(int faltas) { faltasAsistencia.push_back(faltas); }
};

class Profesor {
private:
    string nombre;
    string direccion;
    string numeroTelefono;
    vector<Asignatura*> asignaturasImpartidas;

public:
    Profesor(string _nombre, string _direccion, string _numeroTelefono) : nombre(_nombre), direccion(_direccion), numeroTelefono(_numeroTelefono) {}

    // Métodos para acceder y modificar los atributos
    string getNombre() const { return nombre; }
    string getDireccion() const { return direccion; }
    string getNumeroTelefono() const { return numeroTelefono; }
    void asignarAsignatura(Asignatura* asignatura) { asignaturasImpartidas.push_back(asignatura); }
};

class Asignatura {
private:
    string codigo;
    string nombre;
    string descripcion;
    Profesor* profesorAsociado;

public:
    Asignatura(string _codigo, string _nombre, string _descripcion, Profesor* _profesor) : codigo(_codigo), nombre(_nombre), descripcion(_descripcion), profesorAsociado(_profesor) {}

    // Métodos para acceder a los atributos
    string getCodigo() const { return codigo; }
    string getNombre() const { return nombre; }
    string getDescripcion() const { return descripcion; }
    Profesor* getProfesorAsociado() const { return profesorAsociado; }
};

int main() {
    // Ejemplo de uso
    Estudiante estudiante("Juan", "Calle A, Ciudad B", "2024001");
    Profesor profesor("Dr. Pérez", "Avenida X, Ciudad Y", "123456789");
    Asignatura asignatura("CS101", "Introducción a la Programación", "Curso introductorio a la programación", &profesor);

    // Matricular estudiante en asignatura
    estudiante.matricularAsignatura(&asignatura);

    return 0;
}

File: Trabajo_3/Ejercicios/Ejercicio_Python/Ejercicio3.py
class Estudiante:
    def __init__(self, nombre, direccion, numeroMatricula, asignaturasMatriculadas=None):
        self.nombre = nombre
        self.direccion = direccion
        self.numeroMatricula = numeroMatricula
        if asignaturasMatriculadas is None:
            asignaturasMatriculadas = []
        self.asignaturasMatriculadas = asignaturasMatriculadas

class Profesor:
    def __init__(self, nombre, direccion, telefono, asignaturasImpartidas=None):
        self.nombre = nombre
        self.direccion = direccion
        self.telefono = telefono
        if asignaturasImpartidas is None:
            asignaturasImpartidas = []
        self.asignaturasImpartidas = asignaturasImpartidas

class Asignatura:
    def __init__(self, codigo, nombre, descripcion, profesor, estudiantes=None, notas=None, faltas=None):
        self.codigo = codigo
        self.nombre = nombre
        self.descripcion = descripcion
        self.profesor = profesor
        if estudiantes is None:
            estudiantes = []
        self.estudiantes = estudiantes
        if notas is None:
            notas = []
        self.notas = notas
        if faltas is None:
            faltas = []
        self.faltas = faltas

class Universidad:
    def __init__(self, estudiantes=None, profesores=None, asignaturas=None):
        if estudiantes is None:
            estudiantes = []
        if profesores is None:
            profesores = []
        if asignaturas is None:
            asignaturas = []
        self.estudiantes = estudiantes
        self.profesores = profesores
        self.asignaturas = asignaturas

    def agregarEstudiante(self, estudiante):
        if estudiante not in self.estudiantes:
            self.estudiantes.append(estudiante)
            print("Estudiante agregado.")
        else:
            print("El estudiante ya existe.")

    def agregarProfesor(self, profesor):
        if profesor not in self.profesores:
            self.profesores.append(profesor)
            print("Profesor agregado.")
        else:
            print("El profesor ya existe.")

    def agregarAsignatura(self, asignatura):
        if asignatura not in self.asignaturas:
            self.asignaturas.append(asignatura)
            print("Asignatura agregada.")
        else:
            print("La asignatura ya existe.")

# Creación de una instancia de Universidad
universidad = Universidad()

# Creación de instancias de Estudiante, Profesor y Asignatura
estudiante = Estudiante("Juan", "Calle 123", "123")
profesor = Profesor("Pedro", "Calle 456", "456")
asignatura = Asignatura("123", "Matematicas", "Calculo", profesor)

# Agregar estudiante, profesor y asignatura a la universidad
universidad.agregarEstudiante(estudiante)
universidad.agregarProfesor(profesor)
universidad.agregarAsignatura(asignatura)
